### ISSUE TITLE
Mapreduce is 10x slower than loop.

### ISSUE THREAD

**Issue by oscardssmith** [JuliaLang/julia#38558]
A slack conversation led me to the realization that currently the following two functions have very different speeds.
```
function f(x, y)
    return @inbounds mapreduce(==,+,x, y)
end

function f2(x, y)
	total=0
	@inbounds for i in 1:length(x)
		total += x[i]==y[i]
	end
	return total
end
x = randn(10240), y = similar(x)
@btime f2(x,y)
  1.255 μs (0 allocations: 0 bytes)

@btime f(x,y)
  9.207 μs (1 allocation: 10.19 KiB)
```
This is clearly unfortunate as the `mapreduce` version is much clearer. Is there any way we can make this case for `mapreduce` not have `O(n)` allocation? (or otherwise be faster)

**Comment by jishnub** [comment]
Perhaps this is because `map` allocates an intermediate array? The performance improves noticeably using `MappedArrays`:

```julia
julia> @btime f($x,$y);
  22.352 μs (1 allocation: 10.19 KiB)

julia> @btime f2($x,$y);
  2.178 μs (0 allocations: 0 bytes)

julia> using MappedArrays

julia> function f3(x,y)
           reduce(+, mappedarray(==, x, y))
       end
f3 (generic function with 1 method)

julia> @btime f3($x,$y);
  4.934 μs (0 allocations: 0 bytes)
```

Still not quite as good as the loop, but almost there.

**Comment by timholy** [comment]
If Generators were faster we could just detect the specialization for multiple arrays and use https://github.com/JuliaLang/julia/blob/19b3ec579e8da7b07ae667cd34ab9ede95d4a2a6/base/reduce.jl#L288

Generators are like generalized MappedArrays, but unfortunately they never have quite gotten the same performance. Working on their performance might be the best way to tackle this issue.

**Comment by antoine-levitt** [comment]
> Generators are like generalized MappedArrays, but unfortunately they never have quite gotten the same performance. Working on their performance might be the best way to tackle this issue.

Yes please. Generators are super useful and I use them all the time (much more than higher-order functions).

**Comment by JeffBezanson** [comment]
I get 1.3 us for the loop, 10.4 us for mapreduce, and 7.2 us for the generator version (no allocations). So generators are already faster in this case, but yes need to be faster still.

**Comment by timholy** [comment]
As a reasonable goalpost, here's one more timing comparison:
- `Generator` version: 5.969 μs (0 allocations: 0 bytes)
- [MappedArrays](https://github.com/JuliaArrays/MappedArrays.jl/pull/35) version: 1.631 μs (0 allocations: 0 bytes)
- `f2` version: 1.430 μs (0 allocations: 0 bytes)

**Comment by stevengj** [comment]
`f4(x,y) = mapreduce(((x,y),)->x==y,+,zip(x,y))` also has no allocations and is faster, but it's still not as fast as the loop.

**Comment by oscardssmith** [comment]
That would be a pretty easy to do automatically, right?

**Comment by oscardssmith** [comment]
@timholy is there any good reason why Generators are slow? Is there low hanging fruit? If you have a place to point me, I'd be glad to take a look.

**Comment by mcabbott** [comment]
@wheeheee that's not a bad idea, similar to MappedArrays but in a few lines.

Am not sure what `LinearIndices` is there for; I think that instead you want to reshape in cases where eachindex gives a UnitRange, so that `inds` always has the same size as the first array. Then reductions with `dims` will also work:
```julia
function mapreduce_same_reshape(f, op, A::Vararg{AbstractArray,N}; kw...) where {N}
    tup_f(i) = f(ntuple(j -> @inbounds(A[j][i]), Val(N))...)
    # inds = reshape(eachindex(A...), axes(A[1]))  # this change allows for dims=2 etc
    ei = eachindex(A...)  # when this is CartesianIndices, reshaping it makes it slow, and isn't necc
    inds = ei isa AbstractUnitRange ? reshape(ei, axes(A[1])) : ei
    mapreduce(tup_f, op, inds; kw...)
end

mapreduce_same_reshape(*, +, [1 2; 3 4], [5 6; 7 8]; dims=1) == [26 44]
```

Benchmarks for all the above functions here:

https://gist.github.com/mcabbott/4746e69f321909c3ba209518dc0447bb

**Comment by christiangnrd** [comment]
Found this looking into some other things, seems like performance is now 3x slower instead of 10x. This improvement seems to have come sometime between 1.7 and 1.8.

It also seems like 1.11 brought the allocations up to 3 from 1, and master up to 5.

```julia-repl
julia> using BenchmarkTools; x = randn(10240); y = similar(x);

julia> function f(x, y)
           return @inbounds mapreduce(==,+,x, y)
       end
f (generic function with 1 method)

julia> function f2(x, y)
               total=0
               @inbounds for i in 1:length(x)
                       total += x[i]==y[i]
               end
               return total
       end
f2 (generic function with 1 method)

julia> @btime f2(x,y);
  1.197 μs (0 allocations: 0 bytes)

julia> @btime f(x,y);
  2.956 μs (3 allocations: 10.06 KiB)
```
